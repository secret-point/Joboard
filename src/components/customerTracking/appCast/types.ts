export interface AppCastEventType {
  event: number;
  jid: string; // Job id
  jsid?: string; // Job Seeker Id
}
