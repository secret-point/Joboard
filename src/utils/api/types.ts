import { AxiosError } from "axios";
import {
  ASSESSMENT_ELIGIBILITY_ERROR_CODE,
  CREATE_APPLICATION_ERROR_CODE,
  GET_APPLICATION_ERROR_CODE,
  GET_JOB_INFO_ERROR_CODE,
  GET_NHE_TIME_SLOT_LIST_ERROR_CODE,
  GET_SCHEDULE_DETAIL_ERROR_CODE,
  GET_SCHEDULE_LIST_BY_JOB_ID_ERROR_CODE,
  UPDATE_APPLICATION_ERROR_CODE,
  UPDATE_WORKFLOW_NAME_ERROR_CODE,
  UPDATE_WOTC_STATUS_ERROR_CODE,
  VALIDATE_AMAZON_LOGIN_ID_ERROR_CODE,
  WITHDRAW_APPLICATION_ERROR_CODE
} from "../enums/common";
import { Application, Candidate, Job, NHETimeSlot, PossibleNhePreferenceConfig, Schedule } from "../types/common";

export interface CreateApplicationResponse {
  data: Application;
  error: string;
  errorCode: CREATE_APPLICATION_ERROR_CODE;
}

export interface GetAssessmentElegibilityResponse {
  data: any;
  error: string;
  errorCode: ASSESSMENT_ELIGIBILITY_ERROR_CODE;
}

export interface GetApplicationResponse {
  data: Application;
  error: string;
  errorCode: GET_APPLICATION_ERROR_CODE;
}

export interface WithdrawMultipleApplicationResponse {
  data: Application[];
  error: string;
  errorCode: WITHDRAW_APPLICATION_ERROR_CODE;
}

export interface GetApplicationListResponse {
  data: Application[];
  error: string;
  errorCode: GET_APPLICATION_ERROR_CODE;
}
export interface UpdateApplicationResponse {
  data: Application;
  error: string;
  errorCode: UPDATE_APPLICATION_ERROR_CODE;
}

export interface UpdateWorkflowNameResponse {
  data: Application;
  error: string;
  errorCode: UPDATE_WORKFLOW_NAME_ERROR_CODE;
}

export interface ProxyApiError {
  errorCode: string;
  errorMessage?: string;
  
}
export interface ApiError extends ProxyApiError, AxiosError {}

export interface GetJobInfoResponse {
  data: Job;
  error: string;
  errorCode: GET_JOB_INFO_ERROR_CODE;
}

export interface GetScheduleData {
  availableSchedules: {
    total: number;
    schedules: Schedule[];
  };
  pageFactor: number;
}

export interface GetScheduleListResponse {
  data: GetScheduleData;
  error: string;
  errorCode: GET_SCHEDULE_LIST_BY_JOB_ID_ERROR_CODE;
}

export interface GetNheTimeSlotsDsResponse {
  data: NHETimeSlot[];
  error: string;
  errorCode: GET_NHE_TIME_SLOT_LIST_ERROR_CODE;
}

export interface GetScheduleDetailResponse {
  data: Schedule;
  error: string;
  errorCode: GET_SCHEDULE_DETAIL_ERROR_CODE;
}

export interface GetCandidateResponse {
  data: Candidate;
  error: string;
  errorCode: GET_SCHEDULE_DETAIL_ERROR_CODE;
}

export interface UpdateWotcStatusResponse {
  data: any;
  error: string;
  errorCode: UPDATE_WOTC_STATUS_ERROR_CODE;
}

export interface ValidateamazonLoginIDResponse {
  data: any;
  error: string;
  errorCode: VALIDATE_AMAZON_LOGIN_ID_ERROR_CODE;
}

export interface GetPossibleNhePreferenceResponse {
  data: PossibleNhePreferenceConfig;
  error: string;
  errorCode: any;
}
