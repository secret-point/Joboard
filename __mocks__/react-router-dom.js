// Mock navigate and useParams when needed
// const navigate = jest.fn();
// const useParams = jest.fn();
const useLocation = jest.fn();

module.exports = {
  ...jest.requireActual("react-router-dom"),
  // useNavigate: () => navigate,
  // useParams,
  useLocation,
};
