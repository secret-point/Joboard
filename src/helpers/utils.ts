import isEmpty from "lodash/isEmpty";

export const convertPramsToJson = (params: string) => {
  if (!isEmpty(params)) {
    return JSON.parse(
      '{"' +
        decodeURI(params.substring(1))
          .replace(/"/g, '\\"')
          .replace(/&/g, '","')
          .replace(/=/g, '":"') +
        '"}'
    );
  } else {
    return {};
  }
};

export const launchAuthentication = () => {
  let hash = window.location.hash.substr(1);

  const origin = window.location.origin;
  const redirectUrl = `${origin}/#${hash}`;

  const state = window.reduxStore.getState();
  let url = `${
    state.app.appConfig.authenticationURL
  }/?redirectUrl=${encodeURIComponent(redirectUrl)}`;
  window.location.assign(url);
};

export const isJson = (obj: string) => {
  try {
    JSON.parse(obj);
  } catch (e) {
    return false;
  }
  return true;
};
