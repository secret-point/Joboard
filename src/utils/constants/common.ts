export const HVH_LOCALE = 'hvh-locale';
