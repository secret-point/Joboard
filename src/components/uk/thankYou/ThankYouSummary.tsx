import React, { useEffect } from "react";
import { ButtonVariant } from "@amzn/stencil-react-components/button";
import { Card } from "@amzn/stencil-react-components/card";
import { FlyoutContent, RenderFlyoutFunctionParams, WithFlyout } from "@amzn/stencil-react-components/flyout";
import { IconArrowRight, IconSize } from "@amzn/stencil-react-components/icons";
import { Col, Row } from "@amzn/stencil-react-components/layout";
import { Text } from "@amzn/stencil-react-components/text";
import InnerHTML from "dangerously-set-html-content";
import queryString from "query-string";
import { connect } from "react-redux";
import { useLocation } from "react-router-dom";
import { addMetricForPageLoad } from "../../../actions/AdobeActions/adobeActions";
import { boundGetCandidateInfo } from "../../../actions/CandidateActions/boundCandidateActions";
import { boundGetJobDetail } from "../../../actions/JobActions/boundJobDetailActions";
import { boundGetScheduleDetail } from "../../../actions/ScheduleActions/boundScheduleActions";
import {
  getPageNameFromPath,
  parseQueryParamsArrayToSingleItem,
  redirectToDashboard,
  resetIsPageMetricsUpdated
} from "../../../helpers/utils";
import { ApplicationState } from "../../../reducers/application.reducer";
import { CandidateState } from "../../../reducers/candidate.reducer";
import { JobState } from "../../../reducers/job.reducer";
import { ScheduleState } from "../../../reducers/schedule.reducer";
import { CommonColors } from "../../../utils/colors";
import {
  checkAndBoundGetApplication,
  formatDate,
  getLocale,
  initiateScheduleDetailOnPageLoad
} from "../../../utils/helper";
import { translate as t } from "../../../utils/translator";
import DebouncedButton from "../../common/DebouncedButton";
import { PAGE_ROUTES } from "../../pageRoutes";
import NhePreferenceReview from "../../common/nhe/NhePreferenceReview";

interface MapStateToProps {
  application: ApplicationState;
  candidate: CandidateState;
  schedule: ScheduleState;
  job: JobState;
}

export const ThankYouSummary = (props: MapStateToProps) => {
  const { application, candidate, schedule, job } = props;
  const { search, pathname } = useLocation();
  const pageName = getPageNameFromPath(pathname);
  const queryParams = parseQueryParamsArrayToSingleItem(queryString.parse(search));
  const { applicationId, jobId, scheduleId } = queryParams;
  const applicationData = application.results;
  const { scheduleDetail } = schedule.results;
  const { candidateData, shiftPreferencesData: candidateShiftPreferences } = candidate.results;
  const jobDetail = job.results;
  const nheAppointment = applicationData?.nheAppointment;
  const location = applicationData?.nheAppointment?.location;
  const startTime = nheAppointment?.startTime;
  const endTime = nheAppointment?.endTime;

  useEffect(() => {
    // Refresh and add scheduleId in the url from the jobSelected if it doesn't exist from the query param
    if (!scheduleId && applicationData) {
      initiateScheduleDetailOnPageLoad(applicationData, PAGE_ROUTES.THANK_YOU);
    }
  }, [applicationData]);

  useEffect(() => {
    boundGetCandidateInfo();
  }, []);

  useEffect(() => {
    checkAndBoundGetApplication(applicationId);
  }, [applicationId]);

  useEffect(() => {
    scheduleId && scheduleId!== scheduleDetail?.scheduleId && boundGetScheduleDetail({
      locale: getLocale(),
      scheduleId: scheduleId
    });
  }, [scheduleDetail, scheduleId]);

  useEffect(() => {
    if (jobId) {
      jobId !== jobDetail?.jobId && boundGetJobDetail({ jobId: jobId, locale: getLocale() });
    }
  }, [jobDetail, jobId]);

  useEffect(() => {
    jobDetail && applicationData && candidateData && scheduleDetail && addMetricForPageLoad(pageName);

  }, [jobDetail, applicationData, candidateData, scheduleDetail, pageName]);

  useEffect(() => {
    return () => {
      // reset this so as it can emit new pageload event after being unmounted.
      resetIsPageMetricsUpdated(pageName);
    };
  }, [pageName]);

  const handleGoToDashboard = () => {
    redirectToDashboard();
  };

  const renderNheAppointmentDetails = () => {
    return (
      <>
        <Text fontSize="T300" fontWeight="bold">
          {t("BB-Kondo-ThankYou-nhe-appointment-details-title-text", "Pre-hire appointment details")}
        </Text>

        <Text fontSize="T100">
          {t("BB-Kondo-ThankYou-nhe-appointment-details-date-label", "Date")}: {formatDate(nheAppointment?.dateWithoutFormat, {
            defaultDateFormat: "DD/MM/yyyy",
            displayFormat: "dddd, D MMM"
          })}
        </Text>
        <Text fontSize="T100">
          {t("BB-Kondo-ThankYou-nhe-appointment-details-time-label", "Time")}: {`${startTime} - ${endTime}`}
        </Text>
        <Text fontSize="T100">
          {t("BB-Kondo-ThankYou-nhe-appointment-details-meeting-details-label", "Meeting details")}: {location && `${location.streetAddress} ${location.city} ${location.postalCode}, United Kingdom`}
        </Text>
      </>
    );
  };

  const renderFlyout = ({ close }: RenderFlyoutFunctionParams) => (
    <FlyoutContent
      titleText={t("BB-ThankYou-nhe-appointment-flyout-title-text", "Pre-hire appointment")}
      onCloseButtonClick={close}
    >
      <Col width="100%" padding="S300" gridGap={15}>
        {renderNheAppointmentDetails()}

        <Text fontSize="T300" fontWeight="bold">
          {t("BB-ThankYou-nhe-appointment-flyout-things-to-bring-text", "Things to bring")}
        </Text>
        <ul className="ul-list">
          <li>
            <Text fontSize="T100">
              {t("BB-ThankYou-nhe-appointment-flyout-original-unexpired-photoid-text", "Original unexpired photo ID")}
            </Text>
          </li>
          <li>
            <Text fontSize="T100">
              {t("BB-ThankYou-nhe-appointment-flyout-original-unexpired-work-authorization-document-text", "Original unexpired work authorization document")}
            </Text>
          </li>
        </ul>

        <Text fontSize="T100">
          <InnerHTML html={t("BB-Kondo-ThankYou-nhe-appointment-flyout-acceptable-documents-text", "Click <a href='https://glspub.s3-us-west-2.amazonaws.com/Work+Authorization/UK+Employee+Guide/index.html' target='_blank' rel='noopener noreferrer'>here</a> for a complete list of acceptable documents.")} />
        </Text>

        <Text fontSize="T300" fontWeight="bold">
          {t("BB-ThankYou-nhe-appointment-flyout-special-assistance-text", "Special assistance")}
        </Text>
        <Text fontSize="T100">
          <InnerHTML html={t("BB-Kondo-ThankYou-nhe-appointment-flyout-accommodations-assistance-text", "If you have a disability and consider that you may require an adjustment to Amazon's recruitment process, please contact us to discuss on 0808 164 9427 or by email at myapplication@amazon.co.uk")} />
        </Text>
      </Col>
    </FlyoutContent>
  );

  // temporarily look for either application OR candidate shift preferences at the thank-you summary so candidates
  // who may have submitted with application shift preferences but loaded when candidate shift preferences is toggled
  // on still get the correct experience
  const hasShiftPreferences = applicationData?.shiftPreference ?? candidateShiftPreferences;
  return (
    <Col gridGap="S300" padding={{ top: "S300" }}>

      {/* when shift preference selected and no shift selected, https://sim.amazon.com/issues/Kondo_QA_Issue-51 */}
      { !applicationData?.jobScheduleSelected.scheduleId && hasShiftPreferences && (
        <Col gridGap="S300">
          <Text fontSize="T400">
            {t("BB-Kondo-ThankYou-with-shift-preference-title-text", "Thank you for submitting your shift preferences!")}
          </Text>
          <Text>
            {t("BB-Kondo-ThankYou-with-shift-preference-content-text", "We'll contact you as soon as we find a shift that works for you with your chosen preferences.")}
          </Text>
        </Col>
      )}

      {/* when shift is selected and nhePreference selected */}
      { scheduleDetail && applicationData?.nhePreference && !applicationData?.nheAppointment && (
        <Col gridGap="S300">
          <Text fontSize="T400">
            {t("BB-Kondo-ThankYou-with-nhe-preference-title-text", "Thank you for submitting your pre-hire appointment preferences!")}
          </Text>
          <Text fontSize="T200">
            {t("BB-Kondo-ThankYou-with-nhe-preference-content-text", "We'll contact you as soon as we find a pre-hire appointment that works for you with your chosen preferences.")}
          </Text>
          <NhePreferenceReview nhePreference={applicationData?.nhePreference} />
        </Col>
      )}

      {/* when shift is selected and nhe appointment is selected */}
      { scheduleDetail && applicationData?.nheAppointment && (
        <Col gridGap="S300">
          <Text fontSize="T400">
            {t("BB-ThankYou-thank-you-title-text", "Thank you for applying to Amazon!")}
          </Text>
          <Text fontSize="T200">
            {t("BB-Kondo-ThankYou-description-text", "You will receive an email from us providing further information on the next steps in your application. We look forward to seeing you at your pre-hire appointment.")}
          </Text>

          <Card width="100%" padding="S300" isElevated>
            <Col width="100%" padding="S300" gridGap={15}>
              {renderNheAppointmentDetails()}

              <Row className="border-top-row">
                <WithFlyout renderFlyout={renderFlyout}>
                  {({ open }) => (
                    <Col alignItems="flex-start">
                      <Text
                        color={CommonColors.Blue70}
                        style={{ cursor: "pointer" }}
                        fontWeight="bold"
                        onClick={open}
                      >
                        <Row
                          gridGap={8}
                          alignItems="center"
                        >
                          {t("BB-Kondo-ThankYou-more-details-text", "More details")}
                          <IconArrowRight size={IconSize.ExtraSmall} aria-hidden />
                        </Row>
                      </Text>
                    </Col>
                  )}
                </WithFlyout>
              </Row>
            </Col>
          </Card>
        </Col>
      )}

      <DebouncedButton variant={ButtonVariant.Primary} onClick={handleGoToDashboard}>
        {t("BB-Kondo-ThankYou-return-to-dashboard-text", "Return to dashboard")}
      </DebouncedButton>
    </Col>
  );
};

export const mapStateToProps = (state: MapStateToProps) => {
  return state;
};

export default connect(mapStateToProps)(ThankYouSummary);
