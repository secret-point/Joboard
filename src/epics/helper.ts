import { log, LoggerType } from "../helpers/log-helper";
import { ProxyApiError } from "../utils/api/types";

export const epicSwitchMapHelper = async (response: any) => {
  if (!response.data || !!response.errorCode || !!response.error || response.errorMessage) {
    log("[Epic] epicSwitchMapHelper: Starting to raise proxy api error", response, LoggerType.ERROR);
    const { errorCode } = response;
    throw createProxyApiEpicError(errorCode, response.error || response.errorMessage);
  }
  return response;
};

export const createProxyApiEpicError = (errorCode: string, errorMessage?: string): ProxyApiError => {
  return { errorCode, errorMessage };
};
