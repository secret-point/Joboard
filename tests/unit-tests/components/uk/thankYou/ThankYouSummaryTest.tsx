import * as React from "react";
import { shallow } from "enzyme";
import { useLocation } from "react-router-dom";
import { ThankYouSummary } from "../../../../../src/components/uk/thankYou/ThankYouSummary";
import { CountryCode } from "../../../../../src/utils/enums/common";
import * as helper from "../../../../../src/utils/helper";
import {
  TEST_APPLICATION_ID,
  TEST_APPLICATION_STATE,
  TEST_CANDIDATE_STATE,
  TEST_JOB_ID,
  TEST_JOB_STATE,
  TEST_SCHEDULE_ID,
  TEST_SCHEDULE_STATE,
  TEST_SHIFT_PREFERENCE,
  TestNhePreference
} from "../../../../test-utils/test-data";
import { Application } from "../../../../../src/utils/types/common";

describe("ThankYouSummary", () => {
  const mockLocation = {
    pathname: "/thank-you",
    search: `?jobId=${TEST_JOB_ID}&applicationId=${TEST_APPLICATION_ID}&scheduleId=${TEST_SCHEDULE_ID}`,
    hash: "",
    state: null
  };
  const mockUseLocation = useLocation as jest.Mock;
  mockUseLocation.mockReturnValue(mockLocation);

  // Unit test can't get Katal {{Country}} development value.
  const mockGetCountryCode = jest.spyOn(helper, "getCountryCode");
  mockGetCountryCode.mockReturnValue(CountryCode.US);

  it("should match snapshot - with nhe appointment", () => {
    const shallowWrapper = shallow(
      <ThankYouSummary
        candidate={TEST_CANDIDATE_STATE}
        job={TEST_JOB_STATE}
        application={TEST_APPLICATION_STATE}
        schedule={TEST_SCHEDULE_STATE}
      />);

    expect(shallowWrapper).toMatchSnapshot();
  });

  it("should match snapshot - with nhe Preferences", () => {
    const shallowWrapper = shallow(
      <ThankYouSummary
        candidate={TEST_CANDIDATE_STATE}
        job={TEST_JOB_STATE}
        application={{
          ...TEST_APPLICATION_STATE,
          results: {
            ...TEST_APPLICATION_STATE.results as Application,
            nheAppointment: undefined,
            nhePreference: TestNhePreference
          }
        }}
        schedule={TEST_SCHEDULE_STATE}
      />);

    expect(shallowWrapper).toMatchSnapshot();
  });

  it("should match snapshot - with shift Preferences when application has shift preference", () => {
    const shallowWrapper = shallow(
      <ThankYouSummary
        candidate={TEST_CANDIDATE_STATE}
        job={TEST_JOB_STATE}
        application={{
          ...TEST_APPLICATION_STATE,
          results: {
            ...TEST_APPLICATION_STATE.results as Application,
            nheAppointment: undefined,
            shiftPreference: TEST_SHIFT_PREFERENCE
          }
        }}
        schedule={TEST_SCHEDULE_STATE}
      />);

    expect(shallowWrapper).toMatchSnapshot();
  });

  it("should match snapshot - with shift Preferences when application has no shift, nhe preference and nhe appointment", () => {
    const shallowWrapper = shallow(
      <ThankYouSummary
        candidate={TEST_CANDIDATE_STATE}
        job={TEST_JOB_STATE}
        application={{
          ...TEST_APPLICATION_STATE,
          results: {
            ...TEST_APPLICATION_STATE.results as Application,
            nheAppointment: undefined,
          }
        }}
        schedule={TEST_SCHEDULE_STATE}
      />);

    expect(shallowWrapper).toMatchSnapshot();
  });
});
