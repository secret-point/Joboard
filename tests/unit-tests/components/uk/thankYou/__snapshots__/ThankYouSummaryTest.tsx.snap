// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThankYouSummary should match snapshot 1`] = `
<Col
  gridGap="S300"
  padding={
    Object {
      "top": "S300",
    }
  }
>
  <Col
    gridGap="S300"
  >
    <Text
      font="primary"
      fontSize="T400"
      hyphens="none"
    >
      Thank you for applying to Amazon!
    </Text>
    <Text
      font="primary"
      fontSize="T200"
      hyphens="none"
    >
      You will receive an email from us providing further information on the next steps in your application. We look forward to seeing you at your pre-hire appointment.
    </Text>
    <Card
      backgroundColor="neutral0"
      display="flex"
      isElevated={true}
      padding="S300"
      status="default"
      width="100%"
    >
      <Col
        gridGap={15}
        padding="S300"
        width="100%"
      >
        <Text
          font="primary"
          fontSize="T300"
          fontWeight="bold"
          hyphens="none"
        >
          Pre-hire appointment details
        </Text>
        <Text
          font="primary"
          fontSize="T100"
          hyphens="none"
        >
          Date
          : 
          Thursday, 4 Aug
        </Text>
        <Text
          font="primary"
          fontSize="T100"
          hyphens="none"
        >
          Time
          : 
          13:30
           - 
          14:00
        </Text>
        <Text
          font="primary"
          fontSize="T100"
          hyphens="none"
        >
          Meeting details
          : 
          Onsite - Recruiting Office at Amazon Distribution Center, 3230 International Place Dupont 98327, United Kingdom
        </Text>
        <Row
          className="border-top-row"
        >
          <WithFlyout
            flyoutPosition="trailing"
            renderFlyout={[Function]}
          >
            <Component />
          </WithFlyout>
        </Row>
      </Col>
    </Card>
  </Col>
  <DebouncedButton
    onClick={[Function]}
    variant="primary"
  >
    Return to dashboard
  </DebouncedButton>
</Col>
`;
